src/rust/src/orchard_keys_ffi.rs

```diff
@@ -0,0 +1,294 @@
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_full_viewing_key_eq(
+    k0: *const FullViewingKey,
+    k1: *const FullViewingKey,
+) -> bool {
+    let k0 = unsafe { k0.as_ref() };
+    let k1 = unsafe { k1.as_ref() };
+    k0 == k1
+}
+
+//
+// Spending keys
+//
+
+#[no_mangle]
+pub extern "C" fn orchard_spending_key_for_account(
+    seed: *const u8,
+    seed_len: usize,
+    bip44_coin_type: u32,
+    account_id: u32,
+) -> *mut SpendingKey {
+    let seed = unsafe { slice::from_raw_parts(seed, seed_len) };
+    SpendingKey::from_zip32_seed(seed, bip44_coin_type, account_id)
+        .map(|key| Box::into_raw(Box::new(key)))
+        .unwrap_or(std::ptr::null_mut())
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_spending_key_clone(key: *const SpendingKey) -> *mut SpendingKey {
+    unsafe { key.as_ref() }
+        .map(|key| Box::into_raw(Box::new(*key)))
+        .unwrap_or(std::ptr::null_mut())
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_spending_key_free(key: *mut SpendingKey) {
+    if !key.is_null() {
+        drop(unsafe { Box::from_raw(key) });
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_spending_key_parse(
+    stream: Option<StreamObj>,
+    read_cb: Option<ReadCb>,
+) -> *mut SpendingKey {
+    let mut reader = CppStreamReader::from_raw_parts(stream, read_cb.unwrap());
+
+    let mut buf = [0u8; 32];
+    match reader.read_exact(&mut buf) {
+        Err(e) => {
+            error!(
+                "Stream failure reading bytes of Orchard spending key: {}",
+                e
+            );
+            std::ptr::null_mut()
+        }
+        Ok(()) => {
+            let sk_opt = SpendingKey::from_bytes(buf);
+            if sk_opt.is_some().into() {
+                Box::into_raw(Box::new(sk_opt.unwrap()))
+            } else {
+                error!("Failed to parse Orchard spending key.");
+                std::ptr::null_mut()
+            }
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_spending_key_serialize(
+    key: *const SpendingKey,
+    stream: Option<StreamObj>,
+    write_cb: Option<WriteCb>,
+) -> bool {
+    let key = unsafe { key.as_ref() }.expect("Orchard spending key pointer may not be null.");
+
+    let mut writer = CppStreamWriter::from_raw_parts(stream, write_cb.unwrap());
+    match writer.write_all(key.to_bytes()) {
+        Ok(()) => true,
+        Err(e) => {
+            error!("Stream failure writing Orchard spending key: {}", e);
+            false
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_spending_key_to_full_viewing_key(
+    key: *const SpendingKey,
+) -> *mut FullViewingKey {
+    unsafe { key.as_ref() }
+        .map(|key| Box::into_raw(Box::new(FullViewingKey::from(key))))
+        .unwrap_or(std::ptr::null_mut())
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_spending_key_eq(k0: *const SpendingKey, k1: *const SpendingKey) -> bool {
+    let k0 = unsafe { k0.as_ref() };
+    let k1 = unsafe { k1.as_ref() };
+    k0 == k1
```

Comment:
Hmm, this should not work. I now see that https://github.com/zcash/orchard/pull/149 added `PartialEq` and `Eq` impls for `SpendingKey`, which should not have been allowed. My fault for not reviewing that PR.

We should instead have `impl ConstantTimeEq for SpendingKey`, and then here:
```suggestion
    k0.ct_eq(&k1)
```

Reply:
Or alternatively, we could have had `PartialEq` and `Eq` impls for `SpendingKey`, implemented in terms of a `ConstantTimeEq` impl. That would probably be the more usable approach, but in general we should try to avoid having non-constant-time comparison operators implemented on secret material.

Reply:
I really only added this for use by the tests for the serialization roundtrip, which I could write as `original->serialize-parse->serialize`, comparing the resulting strings, rather than `original->serialize-parse` and comparing the typed spending keys. Should I do that instead?

Reply:
Yes, let's remove `==` and do that instead for the tests here. We still need to fix the `orchard` impl, which I've opened https://github.com/zcash/orchard/issues/241 for.